### PR TITLE
Fix NoClassDefFoundError happening when snakeyaml is used on a custom JRE

### LINKED ISSUE #7580: java.lang.NoClassDefFoundError: java/beans/IntrospectionException if `java.desktop` not included
**Describe the bug**
In version 1.21.0 (this was probably introduced earlier than that), the agent is failing to initialize in the absence of the `java.desktop` module if you use a JDK built with jlink that does not include this module.

**Steps to reproduce**
1. Create a custom JDK distribution using jlink and do not include the `java.desktop` module.
2. Try to initialize the otel agent

**What did you expect to see?**
Initialization works successfully.

**What did you see instead?**
Initialization crashes with the following exception:
```
OpenTelemetry Javaagent failed to start
java.lang.NoClassDefFoundError: java/beans/IntrospectionException
    at org.yaml.snakeyaml.constructor.BaseConstructor.getPropertyUtils(BaseConstructor.java:645)
    at org.yaml.snakeyaml.constructor.BaseConstructor.addTypeDescription(BaseConstructor.java:664)
    at org.yaml.snakeyaml.constructor.Constructor.<init>(Constructor.java:134)
    at org.yaml.snakeyaml.constructor.Constructor.<init>(Constructor.java:79)
    at org.yaml.snakeyaml.constructor.Constructor.<init>(Constructor.java:60)
    at org.yaml.snakeyaml.Yaml.<init>(Yaml.java:64)
    at io.opentelemetry.instrumentation.spring.resources.SpringBootServiceNameDetector.parseNameFromYaml(SpringBootServiceNameDetector.java:178)
    at io.opentelemetry.instrumentation.spring.resources.SpringBootServiceNameDetector.loadFromClasspath(SpringBootServiceNameDetector.java:259)
    at io.opentelemetry.instrumentation.spring.resources.SpringBootServiceNameDetector.findByClasspathApplicationYaml(SpringBootServiceNameDetector.java:158)
    at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
    at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Unknown Source)
    at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(Unknown Source)
    at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(Unknown Source)
    at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
    at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
    at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(Unknown Source)
    at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
    at java.base/java.util.stream.ReferencePipeline.findFirst(Unknown Source)
    at io.opentelemetry.instrumentation.spring.resources.SpringBootServiceNameDetector.createResource(SpringBootServiceNameDetector.java:92)
    at io.opentelemetry.sdk.autoconfigure.ResourceConfiguration.configureResource(ResourceConfiguration.java:59)
```

**What version are you using?**
1.21.0+ (might have been introduced before that)

**Environment**
Compiler: (e.g., "Adoptium 11.0.7")
OS: (e.g., "Alpine Linux")

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Fixes #7580

### CODE REVIEW

File: instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/RuleParser.java [hunk 1]
```diff
@@ -70,6 +70,14 @@ public void addMetricDefsTo(MetricConfiguration conf, InputStream is, String id)
...
       // It contains contextual information about any syntax issues found by the parser.
       logger.log(WARNING, exception.toString());
+    } catch (NoClassDefFoundError e) {
+      // snakeyaml uses java.beans.Introspector to get/set properties on an object; if module
+      // java.desktop is not present, trying to parse yaml will result in a NoClassDefFoundError
+      logger.log(
+          WARNING,
+          "Failed to parse YAML rules from {0}: {1}. "
+              + "If you're using a custom JRE distro built using jlink, ensure that the java.desktop module is included.",
+          new Object[] {id, e.toString()});
```
Comment (trask) [inline instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/RuleParser.java:80]: thoughts on using jackson yaml? cc @jack-berg

also snakeyaml is not very responsive at dealing with CVE reports so [our CVE check](https://github.com/open-telemetry/opentelemetry-java-instrumentation/actions/runs/3907684368/jobs/6677155250) has been failing for a while now due to

```
One or more dependencies were identified with known vulnerabilities in javaagent:

snakeyaml-1.33.jar (pkg:maven/org.yaml/snakeyaml@1.33, cpe:2.3:a:snakeyaml_project:snakeyaml:1.33:*:*:*:*:*:*:*, cpe:2.3:a:yaml_project:yaml:1.33:*:*:*:*:*:*:*) : CVE-2022-1471, CVE-2022-3064, CVE-2021-4235
```

**Review by trask — APPROVED** [review]

File: instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/RuleParser.java (hunk 1 above)
Comment (bcmedeiros) [reply instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/RuleParser.java:80]: I think Jackson would be safer as well.

Comment (laurit) [reply instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/RuleParser.java:80]: https://github.com/FasterXML/jackson-dataformats-text/tree/2.15/yaml seems to use snakeyaml under the hood. Besides [snakeyaml](https://bitbucket.org/snakeyaml/snakeyaml/src/master/) there is also [snakeyaml engine](https://bitbucket.org/snakeyaml/snakeyaml-engine/src/master/) that claims to offer safe parsing of yaml documents.

Comment (mateuszrzeszutek) [reply instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/RuleParser.java:80]: Let me try snakeyaml-engine then.

Comment (mateuszrzeszutek) [reply instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/RuleParser.java]: Done!

File: instrumentation/jmx-metrics/library/src/test/java/io/opentelemetry/instrumentation/jmx/engine/RuleParserTest.java [hunk 2]
```diff
@@ -417,19 +421,6 @@ void testEmptyTagFromParameter() {
     runNegativeTest(yaml);
   }
 
-  @Test
-  void testEmptyPrefix() {
```
Comment (mateuszrzeszutek) [inline instrumentation/jmx-metrics/library/src/test/java/io/opentelemetry/instrumentation/jmx/engine/RuleParserTest.java:421]: This test case no longer makes any sense with the new parser.

File: instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/RuleParser.java [hunk 3]
```diff
@@ -40,9 +46,103 @@ public static RuleParser get() {
 
   private RuleParser() {}
 
-  public JmxConfig loadConfig(InputStream is) throws Exception {
-    Yaml yaml = new Yaml(new Constructor(JmxConfig.class));
-    return yaml.load(is);
+  @SuppressWarnings("unchecked")
+  public JmxConfig loadConfig(InputStream is) {
```
Comment (mateuszrzeszutek) [inline instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/RuleParser.java:50]: I think that the whole JMX config parsing module needs some further refactoring -- we could simplify the code, remove mutability, and clearly separate parsing from converting the config to its "executable" form. Not in this PR though, let's not mix concerns.

**Review by trask — COMMENTED** [review]
@PeterF778 if you have a chance, can you review these changes to the jmx-metrics module?

File: instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/JmxConfig.java [hunk 4]
```diff
@@ -20,22 +20,22 @@ public class JmxConfig {
   //     - JMX_DEFINITION1
   //     - JMX_DEFINITION2
   // The parser is guaranteed to call setRules with a non-null argument, or throw an exception
```
Comment (PeterF778) [inline instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/JmxConfig.java:22]: Please remove this comment line as well, no longer applicable.

**Comment by PeterF778** [comment]
> 

The changes look good. Thanks, @mateuszrzeszutek, for rewriting the JMX Metric Insight yaml parser to work with snakeyaml-engine.

File: instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/JmxConfig.java (hunk 4 above)
Comment (mateuszrzeszutek) [reply instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/JmxConfig.java]: Done. Thanks!

**Pull request merged**